cmd/gcp-controller-manager/app/gke_approver.go

```diff
@@ -482,6 +473,69 @@ func validateTPMAttestation(opts GCPConfig, csr *capi.CertificateSigningRequest,
 	return true, nil
 }
 
+func ekPubAndIDFromCert(opts GCPConfig, blocks map[string]*pem.Block) (*rsa.PublicKey, *nodeidentity.Identity, error) {
+	// When we switch away from ekPubAndIDFromAPI, remove this. Just a
+	// safe-guard against accidental execution.
+	panic("ekPubAndIDFromCert should not be reachable")
+
+	attestCert := blocks["ATTESTATION CERTIFICATE"].Bytes
+	aikCert, err := x509.ParseCertificate(attestCert)
+	if err != nil {
+		return nil, nil, fmt.Errorf("parsing ATTESTATION_CERTIFICATE: %v", err)
+	}
+	if err := opts.TPMEndorsementCACache.verify(aikCert); err != nil {
+		// TODO(awly): handle temporary CA unavailability without denying CSRs.
+		return nil, nil, fmt.Errorf("verifying EK certificate validity: %v", err)
+	}
+	aikPub, ok := aikCert.PublicKey.(*rsa.PublicKey)
+	if !ok {
+		return nil, nil, fmt.Errorf("public key in ATTESTATION CERTIFICATE is %T, want *rsa.PublicKey", aikCert.PublicKey)
+	}
+	nodeID, err := nodeidentity.FromAIKCert(aikCert)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed extracting VM identity from EK certificate: %v", err)
+	}
+
+	return aikPub, &nodeID, nil
+}
+
+func ekPubAndIDFromAPI(opts GCPConfig, blocks map[string]*pem.Block) (*rsa.PublicKey, *nodeidentity.Identity, error) {
+	nodeIDRaw := blocks["VM IDENTITY"].Bytes
+	nodeID := new(nodeidentity.Identity)
+	if err := json.Unmarshal(nodeIDRaw, nodeID); err != nil {
+		return nil, nil, fmt.Errorf("failed parsing VM IDENTITY block: %v", err)
+	}
+
+	srv := betacompute.NewInstancesService(opts.BetaCompute)
+	resp, err := srv.GetShieldedVmIdentity(fmt.Sprint(nodeID.ProjectID), nodeID.Zone, nodeID.Name).Do()
+	if err != nil {
+		if isNotFound(err) {
```

Comment:
Is this actually retry-able? Isn't the GCE API consistent?

Reply:
It's not re-tryable.
If we get a NotFound, regular error is returned which causes CSR deny
If we get any other error, temporaryError is returned which causes CSR ignore and later retry of the approver.